src/halo2/circuit_builder.rs

```diff
@@ -0,0 +1,625 @@
+use crate::number::AbstractNumberType;
+use crate::{analyzer, constant_evaluator, witness_generator};
+use num_traits::{One, Zero};
+
+use super::circuit_data::CircuitData;
+
+fn expression_2_expr(
+    cd: &CircuitData,
+    expr: &Expression,
+    int_to_field: &dyn Fn(&BigInt) -> BigUint,
+) -> (Expr<PlonkVar>, i32) {
+    match expr {
+        Expression::Number(n) => (Expr::Const(int_to_field(n)), 0),
+        Expression::PolynomialReference(polyref) => {
+            assert_eq!(polyref.index, None);
+
+            let rotation = if polyref.next { 1 } else { 0 };
+
+            let plonkvar = PlonkVar::ColumnQuery {
+                column: cd.col(&polyref.name),
+                rotation,
+            };
+
+            (Expr::Var(plonkvar), rotation)
+        }
+        Expression::BinaryOperation(lhe, op, rhe) => {
+            let (lhe, lhe_rot) = expression_2_expr(cd, lhe, int_to_field);
+            let (rhe, rhe_rot) = expression_2_expr(cd, rhe, int_to_field);
+            let res = match op {
+                BinaryOperator::Add => Expr::Sum(vec![lhe, rhe]),
+                BinaryOperator::Sub => Expr::Sum(vec![lhe, Expr::Neg(Box::new(rhe))]),
+                BinaryOperator::Mul => Expr::Mul(vec![lhe, rhe]),
+                _ => unimplemented!("{:?}", expr),
+            };
+            (res, std::cmp::max(lhe_rot, rhe_rot))
+        }
+
+        _ => unimplemented!("{:?}", expr),
+    }
+}
+
+pub(crate) fn analyzed_to_circuit(
+    analyzed: &analyzer::Analyzed,
+    query_callback: Option<impl FnMut(&str) -> Option<AbstractNumberType>>,
+    field_mod: BigUint,
+    verbose: bool,
+    int_to_field: &dyn Fn(&BigInt) -> BigUint,
+) -> PlafH2Circuit {
+    // The structure of the table is as following
+    //
+    // | constant columns | __enable_cur | __enable_next |  witness columns | \
+    // |  bla_bla_bla     |    1         |       1       |   bla_bla_bla    |  |
+    // |  bla_bla_bla     |    1         |       1       |   bla_bla_bla    |  |>  witness + fixed  2^(k-1)
+    // |  ...             |   ...        |      ...      |   ...            |  |
+    // |  bla_bla_bla     |    1         |       0       |   bla_bla_bla    | /     <- __enable_next == 0 since there's no state transition
+    // |  0               |    0         |       0       |   0              | \
+    // |  0               |    0         |       0       |   0              |  |
+    // |  ...             |   ...        |      ...      |   ...            |  |> 2^2-1
+    // |  0               |    0         |       0       |   <unusable>     |  |
+    // |  0               |    0         |       0       |   <unusable>     | /
+
+    // generate fixed and witness (witness).
+
+    let query = |column, rotation| Expr::Var(PlonkVar::ColumnQuery { column, rotation });
+
+    let (fixed, degree) = constant_evaluator::generate(analyzed);
+    let witness = witness_generator::generate(analyzed, degree, &fixed, query_callback, verbose);
```

Comment:
Don't need to change this now, but I would put the halo2 circuit builder conceptually after witness generation, then we don't need to repeat the code here.

Reply:
Where's the code repetition?

Reply:
Calling the witness generator and so on. I think at some point, the backend should be selectable globally and then after we have all constants and witnesses generated through some means, we call the backend.

Reply:
Ah I see. Yea we can extract that later when we have another backend with a similar internal API.

Reply:
What I mean is that there are 4 different ways to run the commandline binary. Many of these generate witness and fixed column output files. All of them should have the option to also generate a halo2 proof.

Reply:
Now I changed halo2 to being a separate step which takes the pil, fixed and witness values.